components/Chips/src/MDCChipField.m

```diff
@@ -0,0 +1,641 @@
+}
+
+- (void)chipTapped:(id)sender {
+  BOOL shouldBecomeFirstResponder = YES;
+  if ([self.delegate respondsToSelector:@selector(chipFieldShouldBecomeFirstResponder:)]) {
+    shouldBecomeFirstResponder = [self.delegate chipFieldShouldBecomeFirstResponder:self];
+  }
+  if (shouldBecomeFirstResponder) {
+    [self becomeFirstResponder];
+  }
+  MDCChipView *chip = (MDCChipView *)sender;
+  if ([self.delegate respondsToSelector:@selector(chipField:didTapChip:)]) {
+    [self.delegate chipField:self didTapChip:chip];
+  }
+}
+
+#pragma mark - MDCChipTextFieldDelegate
+
+- (void)textFieldShouldRespondToDeleteBackward:(UITextField *)textField {
+  if ([self isAnyChipSelected]) {
+    [self removeSelectedChips];
+    [self deselectAllChips];
+  } else {
+    [self selectLastChip];
+  }
+}
+
+#pragma mark - UITextFieldDelegate
+
+- (BOOL)textFieldShouldBeginEditing:(UITextField *)textField {
+  if (textField == self.textField) {
+    [self deselectAllChips];
+  }
+  if ([self.delegate respondsToSelector:@selector(chipFieldDidBeginEditing:)]) {
+    [self.delegate chipFieldDidBeginEditing:self];
+  }
+  return YES;
+}
+
+- (BOOL)textFieldShouldEndEditing:(UITextField *)textField {
+  if ((self.delimiter & MDCChipFieldDelimiterDidEndEditing) == MDCChipFieldDelimiterDidEndEditing) {
+    if (textField == self.textField) {
+      [self commitInput];
+    }
+  }
+  if ([self.delegate respondsToSelector:@selector(chipFieldDidEndEditing:)]) {
+    [self.delegate chipFieldDidEndEditing:self];
+  }
+  return YES;
+}
+
+- (BOOL)textFieldShouldReturn:(UITextField *)textField {
+  BOOL shouldReturn = YES;
+
+  // Chip field content view will handle |chipFieldShouldReturn| if the client is not using chip
+  // field directly. If the client uses chip field directly without the content view and has not
+  // implemented |chipFieldShouldReturn|, then a chip should always be created.
+  if ([self.delegate respondsToSelector:@selector(chipFieldShouldReturn:)]) {
+    shouldReturn = [self.delegate chipFieldShouldReturn:self];
+  }
+  if (shouldReturn) {
+    [self createNewChipWithTextField:textField delimiter:MDCChipFieldDelimiterReturn];
+  }
+
+  return shouldReturn;
+}
+
+- (void)textFieldDidChange {
+  [self deselectAllChips];
+  [self createNewChipWithTextField:self.textField delimiter:MDCChipFieldDelimiterSpace];
+
+  if ([self.delegate respondsToSelector:@selector(chipField:didChangeInput:)]) {
+    [self.delegate chipField:self
+              didChangeInput:[self.textField.text copy]];
+  }
+}
+
+#pragma mark - Private
+
+- (void)removeChipSubview:(MDCChipView *)chip {
+  [chip removeFromSuperview];
+  [chip removeTarget:chip.superview
+              action:@selector(chipTapped:)
+    forControlEvents:UIControlEventTouchUpInside];
+}
+
+- (void)addChipSubview:(MDCChipView *)chip {
+  if (chip.superview != self) {
+    [chip addTarget:self
+             action:@selector(chipTapped:)
+   forControlEvents:UIControlEventTouchUpInside];
+    [self addSubview:chip];
```

Comment:
Should very long chips that extending outside of the chip field's frame be clipped? Is truncating chip text with an ellipsis supported (max chip width)?

Reply:
Fixed.